spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -382,6 +464,31 @@ function onAcknowledgePacket(
 function onTimeoutPacket(packet: Packet) {
   // the packet timed-out, so refund the tokens
   refundTokens(packet)
```

Comment:
I think we need some `if` statement is around `refundTokens`, otherwise if `prevPacket` is not `nil`, aren't we executing refund logic twice?

Alternatively, maybe there is a way in which we can execute `refundTokens` always and if `prevPacket` is not nil, then we execute additional logic in `revertInFlightChanges` (but this logic could be simplified from its current form).

Reply:
i don't believe so. 

So as an example, during a receive in an intermediate chain:

We might mint tokens to `receiverAddrA` and then sending from `receiverAddrA` to the forward escrow account.

The refundTokens will simply send from the forwardEscrowAccount to `receiverAddrA` whereas revertInflightChanges will burn the tokens in `receiverAddrA`'s account.

So the intermediate receive involves a **receive** and then a **send**.

refundTokens reverts the `send`

and `revertInflightPacket` reverts the receive.

Hopefully that's useful context to verify the logic of `revertInflightChanges`

Reply:
I think there are two cases on which executing logic on both `refundTokens` and `revertInflightPacket` will leave the middle chain in an undesirable state in case of an error acknowledgement.

### When middle hop is not source when sending and source when receiving

In `onRecvPacket` vouchers are transferred from escrow account to forwarding account, then in `sendFungibleTokens` vouchers in forwarding account are burnt. In `onAcknowledgePacket` in `refundTokens` vouchers are minted to forwarding account and in `revertInflightPacket` vouchers are minted on escrow account. End state: vouchers in forwarding account.

## When middle hop is not source when sending and not source when receiving

In `onRecvPacket` vouchers are minted to forwarding account, then in `sendFungibleTokens` vouchers in forwarding account are burnt. In `onAcknowledgePacket` in `refundTokens` vouchers are minted to forwarding account and in `revertInflightPacket` nothing is done. End state: vouchers in forwarding account.

I would appreciate if you can take a look at [the diagrams above](https://github.com/cosmos/ibc/pull/1090#pullrequestreview-1981084855), the bottom two diagrams show these two situations, in case I am missing something...

Reply:
Hey guys, in this [PR](https://github.com/cosmos/quint/pull/3) we have the test of all the 4 scenarios depicted above regarding error acknowledgements. Assuming the translation from 'spec' to 'quint' is correct, the specifications have been proven accurate during the execution of error acknowledgments with no undesirable ending state detected. Reviews would be much appreciated. 

For reference I post here  
- [new testing scenario diagrams](https://excalidraw.com/#json=imWPrnjVUYGRX0ruIxjii,bWrrA-vuQx43Sd7Z-pSB2w), adapted to existing nomenclature present in the `ics20.qnt` specification.   
- [the state transition tables for every scenario](https://docs.google.com/spreadsheets/d/1XCI6aNBfqDYhdOpzg0tFxaXwWCPRLaRXgiuPzvePRjk/edit#gid=2106405052)

Potential next steps to ensure the correctness of this spec would be: 
- Adding timeout testing scenarios (even if, since reversion logic is correct, everything should be ok) 
- Add multidenom logic and tests.   

Again thanks @crodriguezvega and @AdityaSripal for the great work!